From cdd54a437f7027a7af24159d3c410a73036d2593 Mon Sep 17 00:00:00 2001
From: Raghuvar S <quic_raghuvar@quicinc.com>
Date: Mon, 17 Oct 2022 09:15:39 +0530
Subject: [PATCH 2/2] Discard .git/ directories with broken symlinks

Recipes with SRC_URI pointing to locally synced
sources having .git/ directories that have broken
symlinks fail in "devtool modify" operation. This
patch discards .git/ directories if they contain
broken symlinks.

Change-Id: Ib9ac95aff86e82945c9d09b9333000d53595c9be
---
 meta/classes/devtool-source.bbclass | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/meta/classes/devtool-source.bbclass b/meta/classes/devtool-source.bbclass
index 41900e651f..9a0e29a4be 100644
--- a/meta/classes/devtool-source.bbclass
+++ b/meta/classes/devtool-source.bbclass
@@ -146,6 +146,17 @@ python devtool_post_unpack() {
                 "doesn't use any source or the correct source "
                 "directory could not be determined" % d.getVar('PN'))
 
+    #Discard .git directories which have files with broken symlinks
+    for root, _, files in os.walk(srcsubdir):
+        if (os.path.basename(root) == ".git"):
+            for f in files:
+                f = os.path.relpath(os.path.join(root,f), srcsubdir)
+                if os.path.islink(os.path.join(srcsubdir, f)):
+                    linkto = os.readlink(os.path.join(srcsubdir, f))
+                    if not os.path.exists(linkto):
+                        shutil.rmtree(os.path.join(root))
+                        break
+
     devbranch = d.getVar('DEVTOOL_DEVBRANCH')
     setup_git_repo(srcsubdir, d.getVar('PV'), devbranch, d=d)
 
-- 
2.17.1

